renderer/html.go

```diff
@@ -4,30 +4,244 @@ import (
+		table.AppendChild(text("\n"))
+	}
+	parent.AppendChild(table)
+	return table
+}
+
+// addColumnGroup adds a columnGroup, if required, to the given table. Cols in the colgroup specify column width and alignment.
+func addColumnGroup(model *tableModel, table *html.Node) {
+	if len(model.request.ColumnFormats) > 0 {
+		colgroup := createNode("colgroup", atom.Colgroup)
+
+		for _, col := range model.columns {
+			node := createNode("col", atom.Col)
+			if len(col.Align) > 0 {
+				setAttribute(node, "class", col.Align)
+			}
+			if len(col.Width) > 0 {
+				setAttribute(node, "style", "width: "+col.Width)
+			}
+			colgroup.AppendChild(node)
+		}
+
+		table.AppendChild(colgroup)
+		table.AppendChild(text("\n"))
+	}
+}
+
+// adds all rows to the table. Rows contain th or td cells as appropriate.
+func addRows(model *tableModel, table *html.Node) {
+	request := model.request
+	for _, row := range request.Data {
+		tr := createNode("tr", atom.Tr)
+		table.AppendChild(tr)
+		for i, col := range row {
+			if model.columns[i].Heading {
+				tr.AppendChild(createNode("th", atom.Th, attr("scope", "row"), parseValue(request, col)))
+			} else {
+				tr.AppendChild(createNode("td", atom.Td, parseValue(request, col)))
+			}
+		}
+		table.AppendChild(text("\n"))
+	}
+}
+
+// addFooter adds a footer to the given element, containing the source and footnotes
+func addFooter(request *models.RenderRequest, parent *html.Node) {
+	footer := createNode("footer", atom.Footer, "\n")
+	if len(request.Source) > 0 {
+		footer.AppendChild(createNode("p", atom.P,
+			attr("class", "table-source"),
+			parseValue(request, "Source: "+request.Source)))
+		footer.AppendChild(text("\n"))
+	}
+	if len(request.Footnotes) > 0 {
+		footer.AppendChild(createNode("p", atom.P,
+			attr("class", "table-notes"),
+			attr("id", "table_"+request.Filename+"_notes"),
+			"Notes"))
+		footer.AppendChild(text("\n"))
+		ol := createNode("ol", atom.Ol, "\n")
+
+		for i, note := range request.Footnotes {
+			ol.AppendChild(createNode("li", atom.Li,
+				attr("id", fmt.Sprintf("table_%s_note_%d", request.Filename, i+1)),
+				parseValue(request, note)))
+			ol.AppendChild(text("\n"))
+		}
+
+		footer.AppendChild(ol)
+		footer.AppendChild(text("\n"))
+	}
+	parent.AppendChild(footer)
+	parent.AppendChild(text("\n"))
+}
+
+// Parses the string to replace \n with <br /> and wrap [1] with a link to the footnote
+func parseValue(request *models.RenderRequest, value string) []*html.Node {
+	hasBr := newLine.MatchString(value)
+	hasFootnote := len(request.Footnotes) > 0 && footnoteLink.MatchString(value)
+	if hasBr || hasFootnote {
+		return replaceValues(request, value, hasBr, hasFootnote)
+	}
+	return []*html.Node{{Type: html.TextNode, Data: value}}
+}
+
+// replaceValues uses regexp to replace new lines and footnotes with <br/> and <a>.../<a> tags, then parses the result into an array of nodes
+func replaceValues(request *models.RenderRequest, value string, hasBr bool, hasFootnote bool) []*html.Node {
```

Comment:
Note that because we have to replace both new lines and any number of footnotes there was no easy way to do this without building a string of html.